### ISSUE TITLE
Windows Service fails to restart

### ISSUE THREAD

**Issue by Crunsher** [Icinga/icinga2#6112]
When receiving a config update, instead of triggering a reload, the service dies.

**Comment by Copis** [comment]
Maybe is not the same problem but in some Window servers, after a config update, the service still up but no longer reports his state.

> Remote Icinga instance 'iamvmcen01' is not connected to 'icinga2'

If I can help with you with logs or something else please let me known

**Comment by dnsmichi** [comment]
This seems to be a follow-up problem from an unreleased patch, git master only.

**Comment by dnsmichi** [comment]
This also happens on Linux, specifically CentOS 7 inside the Vagrant-VM scenario `distributed`.

Seems to come from the change with the reload in #5996

````
commit c418a9611e82dd6694f09b46a88ae478fab3c161
Author: Jean Flach <jean-marcel.flach@icinga.com>
Date:   Wed Jan 17 13:52:23 2018 +0100

    Add systemd watchdog and adjust reload behaviour
```

```
[root@icinga2-satellite1 ~]# icinga2 --version
icinga2 - The Icinga 2 network monitoring daemon (version: v2.8.1-471-g6bb2ed4)

Copyright (c) 2012-2018 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: CentOS Linux
  Platform version: 7 (Core)
  Kernel: Linux
  Kernel version: 3.10.0-693.17.1.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown
```

```
[root@icinga2-satellite1 ~]# systemctl status icinga2
● icinga2.service - Icinga host/service/network monitoring system
   Loaded: loaded (/usr/lib/systemd/system/icinga2.service; enabled; vendor preset: disabled)
   Active: reloading (reload) (Result: exit-code) since Thu 2018-02-22 19:09:02 CET; 1min 12s ago
 Main PID: 13042 (code=exited, status=1/FAILURE)
   CGroup: /system.slice/icinga2.service
```

```
[2018-02-22 19:09:47 +0100] information/ApiListener: New client connection for identity 'icinga2-master1.vagrant.demo.icinga.com' from [192.168.33.101]:35480
[2018-02-22 19:09:47 +0100] information/ApiListener: Requesting new certificate for this Icinga instance from endpoint 'icinga2-master1.vagrant.demo.icinga.com'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Sending config updates for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished sending config file updates for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Syncing runtime objects to endpoint 'icinga2-master1.vagrant.demo.icinga.com'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished syncing runtime objects to endpoint 'icinga2-master1.vagrant.demo.icinga.com'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished sending runtime config updates for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Sending replay log for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished sending replay log for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished syncing endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Applying config update from endpoint 'icinga2-master1.vagrant.demo.icinga.com' of zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//.timestamp
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/additional_services.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/api-users.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/bp.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/commands.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/cube.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/groups.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/hosts.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/many.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/maps.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/notifications.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/services.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/templates.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/timeperiods.conf
...skipping...
[2018-02-22 19:09:47 +0100] information/ApiListener: Sending replay log for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished sending replay log for endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Finished syncing endpoint 'icinga2-master1.vagrant.demo.icinga.com' in zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Applying config update from endpoint 'icinga2-master1.vagrant.demo.icinga.com' of zone 'master'.
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//.timestamp
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/additional_services.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/api-users.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/bp.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/commands.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/cube.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/groups.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/hosts.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/many.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/maps.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/notifications.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/services.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/templates.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/timeperiods.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Updating configuration file: /var/lib/icinga2/api/zones/satellite//_etc/users.conf
[2018-02-22 19:09:47 +0100] information/ApiListener: Applying configuration file update for path '/var/lib/icinga2/api/zones/satellite' (24319 Bytes). Received timestamp '2018-02-22 18:44:47 +0100' (1519321487.854484), Current timestamp '1970-01-01 01:00:00 +0100' (0.000000).
[2018-02-22 19:09:47 +0100] information/ApiListener: Restarting after configuration change.
[2018-02-22 19:09:49 +0100] information/Application: Got reload command: Starting new instance.
[2018-02-22 19:09:49 +0100] information/Application: Reload requested, letting new process take over.
[2018-02-22 19:09:49 +0100] information/NotificationComponent: 'notification' started.
[2018-02-22 19:09:49 +0100] information/ApiListener: 'api' started.
[2018-02-22 19:09:49 +0100] information/ApiListener: Adding new listener on port '5665'
[2018-02-22 19:09:49 +0100] critical/TcpSocket: Invalid socket: Address already in use
Context:
        (0) Activating object 'api' of type 'ApiListener'

[2018-02-22 19:09:49 +0100] critical/ApiListener: Cannot bind TCP socket for host '' on port '5665'.
Context:
        (0) Activating object 'api' of type 'ApiListener'

[2018-02-22 19:09:49 +0100] critical/ApiListener: Cannot add listener on host '' for port '5665'.
Context:
        (0) Activating object 'api' of type 'ApiListener'
```

**Comment by Crunsher** [comment]
I could not reproduce this with the vagrant-distributed setup. But I noticed we don't hand over the syslog :/ journalctl stops writing after a reload signal (sd_notify should take care of that?!)

**Comment by winter1967** [comment]
W've over 500 Windows Server, were i rolled out the new Icinga2.8.1 Agent. Mostly, the serivce runs without errors, but on all Windows 2008 Exchange Servers (don't know if the system or the high load matters), the icinga2 service stops after a while with the error, that have been listed above.``

**Comment by Crunsher** [comment]
@winter1967 This is a different issue, this here is about the reload failing after receiving new config and starting again. Yours sounds more like the trouble we have with tcp connections not being closed properly on Windows. Yet I haven't seen that happening to a running system

**Comment by winter1967** [comment]
Because a windows server receiving every hours new configs, it's restarts sometimes at a day. So at my servers, the issue happend after restart, too. Maybe my description misleading, here is the whole errortext:
`[2018-02-26 15:06:15 +0100] information/ApiListener: Applying config update from endpoint 'icingaServer.domain.name.de' of zone 'master'.
[2018-02-26 15:06:15 +0100] information/ApiListener: Updating configuration file: C:\ProgramData\icinga2\var/lib/icinga2/api/zones/director-global//.timestamp
[2018-02-26 15:06:15 +0100] information/ApiListener: Updating configuration file: C:\ProgramData\icinga2\var/lib/icinga2/api/zones/director-global//director/service_templates.conf
[2018-02-26 15:06:15 +0100] information/ApiListener: Applying configuration file update for path 'C:\ProgramData\icinga2\var/lib/icinga2/api/zones/director-global' (4564 Bytes). Received timestamp '2018-02-26 15:06:07 +0100' (1519653967.694229), Current timestamp '2018-02-26 14:59:49 +0100' (1519653589.131403).
[2018-02-26 15:06:15 +0100] warning/ApiListener: Ignoring config update for unknown zone 'global-templates'.
[2018-02-26 15:06:15 +0100] information/ApiListener: Restarting after configuration change.
[2018-02-26 15:06:16 +0100] information/Application: Got reload command: Starting new instance.
[2018-02-26 15:06:20 +0100] information/ApiListener: Adding new listener on port '5665'
[2018-02-26 15:06:20 +0100] critical/TcpSocket: Invalid socket: 10048, "Only one usage of each socket address (protocol/network address/port) is normally permitted."
Context:
	(0) Activating object 'api' of type 'ApiListener'

[2018-02-26 15:06:20 +0100] critical/ApiListener: Cannot bind TCP socket for host '::' on port '5665'.
Context:
	(0) Activating object 'api' of type 'ApiListener'

[2018-02-26 15:06:20 +0100] critical/ApiListener: Cannot add listener on host '::' for port '5665'.`

**Comment by Crunsher** [comment]
Ahhhh, this should get fixed with  #6116

**Comment by hijacker7** [comment]
I revert to Agent Version 2.7.2 and it's working again. 
The step between (v2.8.0) has the same problem like v2.8.1.

**Comment by dnsmichi** [comment]
I've tested the fix with #6166, this works for me.

**Comment by dnsmichi** [comment]
Not relevant anymore.
